crates/oxc_linter/src/rules/oxc/no_redundant_constructor_init.rs

```diff
@@ -0,0 +1,164 @@
+use oxc_ast::{
+    AstKind,
+    ast::{AssignmentTarget, Expression, MethodDefinitionKind, Statement},
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{AstNode, context::LintContext, rule::Rule};
+
+fn no_redundant_constructor_init_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Explicit initialization of public members is redundant")
+        .with_help("Remove the explicit initialization")
+        .with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoRedundantConstructorInit;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Prevents redundant initialization of class members within a constructor.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Arguments marked as `public` within a constructor are automatically initialized.
+    /// Providing an explicit initialization is redundant and can be removed.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// class Foo {
+    ///   constructor(public name: unknown) {
+    ///     this.name = name;
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// class Foo {
+    ///   constructor(public name: unknown) {}
+    /// }
+    /// ```
+    NoRedundantConstructorInit,
+    oxc,
+    correctness,
+    pending,
+);
+
+impl Rule for NoRedundantConstructorInit {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::MethodDefinition(method) = node.kind() else {
+            return;
+        };
+        if method.kind != MethodDefinitionKind::Constructor {
+            return;
+        }
+        let public_members = method
+            .value
+            .params
+            .items
+            .iter()
+            .filter_map(
+                |param| if param.is_public() { param.pattern.get_identifier_name() } else { None },
+            )
+            .collect::<Vec<_>>();
+        if public_members.is_empty() {
+            return;
+        }
+        let Some(body) = method.value.body.as_ref() else {
+            return;
+        };
+        body.statements.iter().for_each(|stmt| {
```

Comment:
probably not for now, but i wonder whether it'd be better to grab the public members references and check that way. WIth this approach, if you had it assigned in a block, it wouldn't report an err

Reply:
Nice, didn’t know that was available. Will definitely improve in a follow up.